### PR TITLE
Feature/2363 respecification of wiki page delete

### PR DESCRIPTION
**Pull request by tilt**
https://www.openproject.org/issues/2363

### CODE REVIEW

**Comment by mfrister** [comment]
For the next time: Keeping the length of the first line of commit messages <= 72 makes GitHub always show it as a whole, so I (and I think others too) try to do that. Below an empty line, i.e. line 3, you can write as much as you like.

Waiting for the tests to finish, then taking a look at it.

File: app/assets/stylesheets/default/application.css.erb [hunk 1]
```diff
@@ -2975,6 +2975,9 @@ form.wiki_menu_item_form p.wiki_menu_item_optional_links {
   padding-top: 0px;
 }
 
+#wiki-menu-item-new-main-item-explanation {
```
Comment (mfrister) [inline app/assets/stylesheets/default/application.css.erb]: Do we really need such a specific id here? We probably already have a more generic way of displaying such an explanation. @cratz  should know more about this.

File: config/locales/en.yml [hunk 2]
```diff
@@ -1528,6 +1529,7 @@ en:
   wiki_menu_item: "Menu item"
   wiki_menu_item_for: "Menu item for wikipage \"%{title}\""
   wiki_menu_item_setting: "Visibility"
+  wiki_menu_item_new_main_item_explanation: "You are deleting the only main item. You can now choose a wiki page for which a new main item will be generated."
```
Comment (mfrister) [inline config/locales/en.yml]: I think 'wiki' and 'menu' should be within these sentences, e.g.:
'... deleting the only Wiki menu item. You can now choose a new Wiki page for which a new main item will be generated'

File: spec/controllers/wiki_menu_items_controller_spec.rb [hunk 3]
```diff
@@ -101,13 +101,13 @@
...
 
-    before { get :select_main_menu_item, project_id: project, id: wiki_page.title }
+    before { get :select_main_menu_item, project_id: project, id: wiki_page.id }
     subject { assigns['possible_wiki_pages'] }
 
     context 'when selecting a new wiki page to replace the current main menu item' do
       it { should include another_wiki_page }
       it { should include child_page }
-      it { should_not include wiki_page }
+      it { should include wiki_page }
```
Comment (mfrister) [inline spec/controllers/wiki_menu_items_controller_spec.rb]: Also testing for a child of `wiki_page` would be great as this was the problem I found with the previous implementation.

`child_page` apparently is a child of `another_wiki_page`.

File: spec/models/wiki_page_spec.rb [hunk 4]
```diff
@@ -48,40 +47,32 @@
...
-        grand_child_page.nearest_parent_menu_item(is_main_item: true).should == wiki_page_wiki_menu_item
+        grand_child_page.nearest_parent_menu_item(is_main_item: true).should == wiki_page.menu_item
       end
     end
   end
 
   describe '#destroy' do
     context 'when the only wiki page is destroyed' do
-      before do
+      before :each do
```
Comment (mfrister) [inline spec/models/wiki_page_spec.rb:57]: `before` and `before :each` should be the same: http://stackoverflow.com/questions/12152342/before-hook-in-rspec-same-as-before-all :)

File: spec/models/wiki_spec.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+require 'spec_helper'
+
+describe WikiPage do
```
Comment (mfrister) [inline spec/models/wiki_spec.rb]: Shouldn't this be `Wiki`?

Comment (tilt) [inline spec/models/wiki_spec.rb]: Right!

**Comment by mfrister** [comment]
Deleting the last wiki page and then having the same state as when visiting a project wiki for the first time works :+1: 

When removing the last Wiki page from the main menu, I'm redirected to the "Select new menu item" page, but there I select an "empty" first item, which results in no menu item being shown. I'd consider this a bug or is this a feature I don't understand?

File: app/assets/stylesheets/default/application.css.erb (hunk 1 above)
Comment (tilt) [inline app/assets/stylesheets/default/application.css.erb]: Yes, I think there might be a rule in application.css. Unfortunately we haven't separated the css by concerns yet.

**Comment by tilt** [comment]
@meeee all changes applied

**Comment by mfrister** [comment]
:+1:

**Pull request merged**